plugins/pay.c

```diff
@@ -2003,31 +2004,45 @@ static struct command_result *json_paymod(struct command *cmd,
 	p = payment_new(cmd, cmd, NULL /* No parent */, paymod_mods);
 	p->invstring = tal_steal(p, b11str);
 
-	b11 = bolt11_decode(cmd, b11str, plugin_feature_set(cmd->plugin),
-			    NULL, chainparams, &fail);
-	if (b11) {
+	is_b12 = strlen(b11str) > 3 &&
+		 (strstarts(b11str, "lni") || strstarts(b11str, "lno") ||
+		  strstarts(b11str, "lnr"));
```

Comment:
Hmm, I think it's clearer to use two fail ptrs (esp since this test doesn't work correctly for upper case, and I worry about it breaking if we allow lightning:// prefixes etc)?

Then prefer to the bolt11 failstr if both fail.  Later we can apply some kind of heuristic like this if necessary (maybe create a helper like looks_like_bolt12_invoice())?

Reply:
Untested:

```diff
diff --git a/common/bolt12.c b/common/bolt12.c
index 9cff0ab88..174c66430 100644
--- a/common/bolt12.c
+++ b/common/bolt12.c
@@ -289,6 +289,11 @@ struct tlv_invoice *invoice_decode_nosig(const tal_t *ctx,
 	return invoice;
 }
 
+bool smells_like_bolt12_invoice(const char *str)
+{
+	return strstarts(str, "lni1") || strstarts(str, "LNI1");
+}
+
 static void add_days(struct tm *tm, u32 number)
 {
 	tm->tm_mday += number;
diff --git a/common/bolt12.h b/common/bolt12.h
index 797c5b64b..e9edd9da2 100644
--- a/common/bolt12.h
+++ b/common/bolt12.h
@@ -99,6 +99,9 @@ struct tlv_invoice *invoice_decode_nosig(const tal_t *ctx,
 					 const struct chainparams *must_be_chain,
 					 char **fail);
 
+/* For better error msgs when we get a bolt11 or bolt12. */
+bool smells_like_bolt12_invoice(const char *str);
+
 /* Check a bolt12-style signature. */
 bool bolt12_check_signature(const struct tlv_field *fields,
 			    const char *messagename,
```